src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/GeneratedCodeContainer.cs

```diff
@@ -133,10 +233,14 @@ public void SetText(SourceText sourceText)
                     throw new ArgumentNullException(nameof(sourceText));
                 }
 
-                var e = new TextChangeEventArgs(_currentText, sourceText);
-                _currentText = sourceText;
+                lock (_outerLock)
+                {
 
-                TextChanged?.Invoke(this, e);
+                    var e = new TextChangeEventArgs(_currentText, sourceText);
+                    _currentText = sourceText;
+
+                    TextChanged?.Invoke(this, e);
```

Comment:
So this can now be raised on any thread? Is that allowed?

Reply:
It kind of has to be unless we make `HostDocument`s dependent on a foreground dispatcher which would be a semi-breaking change and then async void dispatches an event trigger. Seemed a bit questionable so I didn't entertain the idea.

What do you think?

Reply:
Note that this would also mean the HostDocument takes in a dispatcher in its constructor since the dispatcher is a MEF component.

Reply:
yeah, so I'm asking is that allowed - I genuinely don't know the answer